lib/hexstring.ml

```diff
@@ -1,17 +1,35 @@
 (* encoding bytearray -> hexstring *)
 
 let encode (bytearray : bytes) : string =
-  let rec encode_rec fold bytearray =
-    let length = Bytes.length bytearray in
-    if length = 0 then fold
+  let dummy_char = '_' in
+  let start_of_digit_0_in_ascii_table = 0x30 in
+  let start_of_lower_case_a_in_ascii_table = 0x61 in
+  let hex_digit_of_int (x : int) : char =
+    assert (x >= 0);
+    assert (x < 16);
+    char_of_int (
+      if x < 10 then
+        x + start_of_digit_0_in_ascii_table
+      else
+        x + start_of_lower_case_a_in_ascii_table
```

Comment:
shouldn't this be `x - 10 + start_of_lower_case_a_in_ascii_table`

Reply:
mind adding a test to catch this :) ? maybe we can modify the encoding test to:

```ocaml
let%test "encoding" =
  let bytearray = Bytes.of_string "\x01\x02\x03\x04\x0a\xbd\0xff" in
  let hexstring = encode bytearray in
  hexstring = "010203040abdff"
```

Reply:
Will add.

And I also think this calls for a property based testing suite, which I'll add shortly...

Reply:
> And I also think this calls for a property based testing suite, which I'll add shortly...

Right you already got one in `lib/tests`, I'm being silly : D